### PR TITLE
Custom Auth: add requestInterceptor for custom x-* request headers

### PR DESCRIPTION
**Pull request by shenj**
## Summary

Adds a new `requestInterceptor` option to `CustomAuthOptions` that lets applications attach additional `x-*` headers to every custom-auth backend request (sign-in, sign-up, reset-password, register).

The primary use case is integrating with third-party fraud / bot-detection SDKs that require vendor-supplied headers on the auth endpoints.

## API

```ts
import { CustomAuthRequestInterceptor } from "@azure/msal-browser/custom-auth";

const interceptor: CustomAuthRequestInterceptor = {
    addAdditionalHeaderFields(requestUrl: URL) {
        if (requestUrl.pathname.includes("/oauth2/v2.0/initiate")) {
            return {
                value_1: "customer_header_1",          // Ignored: no "x-" prefix
                "x-client-header": "customer_header_2",  // Ignored: reserved prefix
                "X-my-custom-header": "my data",         // Sent
            };
        }
        return null;
    },
};

const config: CustomAuthConfiguration = {
    auth: { clientId, authority },
    customAuth: {
        authApiProxyUrl,
        requestInterceptor: interceptor,
    },
};
```

The interceptor may return either a synchronous record or a `Promise`.

## Filter rules (same as iOS doc)

* Names must start with `x-` (case-insensitive); others are dropped.
* Reserved prefixes are dropped: `x-client-`, `x-ms-`, `x-broker-`, `x-app-`.
* User headers that pass the filter take precedence over MSAL's common headers.
* Exceptions thrown by the interceptor are captured and logged via `Logger.warningPii`; the request still goes through so user-supplied code cannot break authentication.

## Scope

* Filtering applied only to custom-auth backend requests (anything routed through `BaseApiClient.request`).
* Wiring: `CustomAuthStandardController` → `CustomAuthApiClient` → each sub-client (`SignInApiClient`, `SignupApiClient`, `ResetPasswordApiClient`, `RegisterApiClient`) → `BaseApiClient`.
* No cache schema impact.

## Tests

* `CustomHeaderUtils.spec.ts` — 16 cases covering each filter rule, casing, and edge cases.
* `BaseApiClientInterceptor.spec.ts` — 8 cases covering: no-interceptor, URL passed to interceptor, filtering, precedence, sync return, async return, null return, sync throw, async rejection (all swallowed).
* `CustomAuthApiClient.spec.ts` extended with 2 construction cases.
* Full `lib/msal-browser` suite: **1643 tests pass**; lint, format, build, and api-extractor all clean.

## Changefile

`change/@azure-msal-browser-ec2f3299-….json` — `minor` (new public option).

### CODE REVIEW

File: change/@azure-msal-browser-ec2f3299-3059-4820-b675-ab278d78e0ae.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Custom Auth: add `requestInterceptor` configuration option that lets apps attach additional `x-*` headers to custom-auth backend requests (e.g., for fraud/bot-detection vendors). Headers without the `x-` prefix and headers starting with reserved prefixes (`x-client-`, `x-ms-`, `x-broker-`, `x-app-`) are filtered out.",
```
[the remark this answers is not in the document]
Comment (shenj) [reply change/@azure-msal-browser-ec2f3299-3059-4820-b675-ab278d78e0ae.json]: Done — added `[#8587](https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/8587)` to the changefile comment in 864c6c697.

**Review by antrix1989 — DISMISSED** [review]

**Review by yongdiw — DISMISSED** [review]

File: lib/msal-browser/src/custom_auth/core/utils/CustomHeaderUtils.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        if (!lowerName.startsWith(CustomHeaderConstants.REQUIRED_PREFIX)) {
+            continue;
+        }
+
+        const startsWithReservedPrefix =
+            CustomHeaderConstants.RESERVED_PREFIXES.some((prefix) =>
+                lowerName.startsWith(prefix)
+            );
+
+        if (startsWithReservedPrefix) {
```
Comment (mustafamizrak) [inline lib/msal-browser/src/custom_auth/core/utils/CustomHeaderUtils.ts]: Is it same behavior in iOS side to overwrite existing headers?

File: lib/msal-browser/test/custom_auth/core/utils/CustomHeaderUtils.spec.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        const result = filterCustomHeaders({
+            "": "empty-name",
+            "   ": "whitespace-name",
+            "x-keep": "value",
+        });
+
+        expect(result).toEqual({ "x-keep": "value" });
+    });
+
+    it("trims whitespace around header names before applying rules", () => {
```
Comment (mustafamizrak) [inline lib/msal-browser/test/custom_auth/core/utils/CustomHeaderUtils.spec.ts:154]: no test for case insensitivity. i.e. `x-Custom-header` and `x-custom-Header` or `x-custom-header.`

File: lib/msal-browser/test/custom_auth/core/network_client/custom_auth_api/BaseApiClientInterceptor.spec.ts [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        });
+
+        const headers = lastSentHeaders();
+        expect(headers[AADServerParamKeys.CLIENT_REQUEST_ID]).toBe("corr");
+        // Confirm no user-vendor header (i.e., x-* not starting with reserved prefixes)
+        // is present when no interceptor is configured.
+        const customHeaderKeys = Object.keys(headers).filter((k) => {
+            const lk = k.toLowerCase();
+            return (
+                lk.startsWith("x-") &&
```
Comment (mustafamizrak) [inline lib/msal-browser/test/custom_auth/core/network_client/custom_auth_api/BaseApiClientInterceptor.spec.ts]: <NIT> Duplicated usages as defined in `CustomHeaderConstants.ts`.

File: lib/msal-browser/src/custom_auth/core/network_client/custom_auth_api/BaseApiClient.ts [hunk 5]
```diff
@@ -173,4 +188,28 @@ export abstract class BaseApiClient {
             responseError.timestamp
         );
     }
+
```
Comment (mustafamizrak) [inline lib/msal-browser/src/custom_auth/core/network_client/custom_auth_api/BaseApiClient.ts:191]: This method warn messages like here https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/3112/changes/BASE..1611df94548697bcbb0ff0e67d0357e2b833a3d4#diff-2a3bf1f0fb795803ab6b829ae9a6db04d93821ad115b4734f366650dbb17c35aR129 for consistency?

File: lib/msal-browser/test/custom_auth/core/network_client/custom_auth_api/BaseApiClientInterceptor.spec.ts (hunk 4 above)
Comment (shenj) [reply lib/msal-browser/test/custom_auth/core/network_client/custom_auth_api/BaseApiClientInterceptor.spec.ts]: Done in 9798b034a — replaced the duplicated `x-`/`x-client-`/`x-ms-`/`x-broker-`/`x-app-` literals in the spec with an `isUserCustomHeader` helper backed by `CustomHeaderConstants.REQUIRED_PREFIX` and `CustomHeaderConstants.RESERVED_PREFIXES`.

File: lib/msal-browser/src/custom_auth/core/utils/CustomHeaderUtils.ts (hunk 2 above)
Comment (shenj) [reply lib/msal-browser/src/custom_auth/core/utils/CustomHeaderUtils.ts]: Yes — same behavior as iOS. The iOS doc states *"If a header you provide has the same name as one of MSAL's own internal headers, your value takes precedence"*, which is what our `{ ...commonHeaders, ...additionalHeaders }` merge does in `BaseApiClient.request`. In practice MSAL's common headers all start with reserved prefixes (`x-client-*`, `x-ms-*`) or are `Content-Type`, so they're either filtered out before the merge or wouldn't collide. The interface JSDoc already documents this precedence rule. Reference: [iOS tutorial — header naming rules](https://learn.microsoft.com/en-us/entra/identity-platform/tutorial-native-authentication-ios-swift-custom-headers#understand-header-naming-rules).

File: lib/msal-browser/test/custom_auth/core/utils/CustomHeaderUtils.spec.ts (hunk 3 above)
Comment (shenj) [reply lib/msal-browser/test/custom_auth/core/utils/CustomHeaderUtils.spec.ts:154]: Added in 9798b034a — three new tests in `CustomHeaderUtils.spec.ts`:

1. `keeps differently-cased variants of the same name as distinct entries (case-sensitive keys, case-insensitive prefix match)` — `x-Custom-header`, `x-custom-Header`, `X-CUSTOM-HEADER` are all preserved as separate `Record` entries (JS object keys are case-sensitive) while still passing the case-insensitive `x-` prefix check.
2. `treats reserved-prefix checks case-insensitively across casings of the same logical name` — `x-Client-Header`, `X-CLIENT-header`, `x-client-Header` are all dropped.
3. The new warning-emission tests also exercise mixed casings (`X-MS-Custom`, etc.) to cover the case-insensitive reserved-prefix path.

File: lib/msal-browser/src/custom_auth/core/network_client/custom_auth_api/BaseApiClient.ts (hunk 5 above)
Comment (shenj) [reply lib/msal-browser/src/custom_auth/core/network_client/custom_auth_api/BaseApiClient.ts:191]: Done in 9798b034a — `filterCustomHeaders` now takes an optional `Logger` + `correlationId` and emits a PII-safe warning (`Logger.warningPii`) for each dropped header, naming the field and the reason, matching the Android `NativeAuthHeaderValidator` message style:

- `Additional header field "<name>" must start with the "x-" prefix. Ignoring.`
- `Additional header field "<name>" uses reserved prefix "<prefix>". Ignoring.`

`BaseApiClient.getAdditionalHeaders` forwards its logger and correlationId into the filter. Two new tests in `CustomHeaderUtils.spec.ts` assert both warning paths and that kept headers don't emit warnings.

**Review by mustafamizrak — APPROVED** [review]

**Review by yongdiw — APPROVED** [review]
